metagraph/core/compiler.py

```diff
@@ -22,13 +22,17 @@ class DaskSubgraph:
 
 
 def extract_compilable_subgraphs(
-    dsk: Dict, compiler: str, include_singletons=True
+    dsk: Dict, compiler: str, output_keys: List[str], include_singletons=True
```

Comment:
My understanding is that the output keys should always be a subset of the nodes without children in the DAG. If this is not correct, then disregard these comments. 

Would it be worth having a default of `output_keys=None` so that we can automatically find the output keys? I'm imagining two ways of implementing this:
* Find the output keys up front via something like:
```
if output_keys is None:
    output_keys = [key for key in dsk.keys() if len(dependendents[key]) == 0]
```
* Replace the code near lines 71-76 with 
```
        if (
            len(next_key_dependencies) == 1
            and len(key_dependents) == 1
            and next_key in key_dependents
            and key in next_key_dependencies
        ):
            if (output_keys is None and len(depenents[next_key]) > 0) or (next_key not in output_keys):
                ...
```

Reply:
Should we also sanity check that the output keys is a subset of the nodes without children in the DAG? e.g.
```
for key in dsk.keys():
    if len(dependents[key]) == 0 and key not in output_keys:
        raise ValueError(f'{key} is an output key but is not specified to be.')
```

Reply:
If my understanding that the output keys should always be a subset of the nodes without children in the DAG is incorrect, then should we prune nodes with no children/dependencies that are not in `output_keys`?

For example, should we prune this graph (where C is not an output key and E is an output key)
```
               A
              / \
             B   D
             |   |
	     C   E
```
into `A -> D -> E`?

Reply:
This is a good point.  The output keys list ultimately comes from Dask when the user calls something like `compute(a,b,c)` (in which case it will pass a, b, and c's keys to the optimize method) .  For other collections, Dask calls a `cull()` method to remove unneeded keys from the graph: `https://github.com/dask/dask/blob/master/dask/optimization.py#L18-L61`

Since Dask doesn't call `cull()` for every collection, I'll add it to Metagraph's optimize method before compiling subgraphs, which is a good idea regardless of whether we can compile nodes.